### PR TITLE
[HttpKernel] Use VarDumper in the profiler

### PR DESCRIPTION
**Pull request by nicolas-grekas**
Some changes on top of https://github.com/symfony/symfony/pull/19614 for @wouterj 
The first commit is exactly yours, the second one is my contribution.
If you're fine with it, please update your PR with this patch, doing e.g.:

```
git fetch https://github.com/nicolas-grekas/symfony.git profiler_dumper
git push -f origin FETCH_HEAD:profiler_dumper
```

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
@wouterj I rewrote the git history so that we can split the work on the logger panel in another PR.
I think you can reset --hard to the first commit here, keep working on it on your onw if you'd like to,
and I will open a second PR with my commit (the 2nd) later on.
WDYT?

**Comment by wouterj** [comment]
@nicolas-grekas I just started working on it again yesterday afternoon. I think the logger panel is finished now? (readding the hide/show buttons for the context fixed the load problem). If it's finished, I think we can include it in the PR directly. Otherwise, I agree with splitting it.

**Comment by nicolas-grekas** [comment]
I didn't implement the hide/show button yet.
Btw, I just spotted an issue in the current first commit (the dump headers should be resetted when parsing a new twig template). I'm on it, I'll keep you posted here.

**Comment by wouterj** [comment]
> I didn't implement the hide/show button yet.

I did locally (for both log context and deprecation trace). Btw, I don't like the deprecation trace that much, it's not very readable imo.

**Comment by nicolas-grekas** [comment]
OK great, push it on your branch, I'll integrate it here. About the deprecation trace, it's the generic output for traces. I think it's important to have the same way to display traces everywhere. If you have any suggestions to make it better, I'd be happy to listen to them.

**Comment by wouterj** [comment]
Here you go: https://github.com/wouterj/symfony/commit/276a3bbc82aab80221387339dcaeca5ee093285c

> If you have any suggestions to make it better, I'd be happy to listen to them.

I think there are 2 things that confused me most:
- It uses the purple and yellow colors, which don't look nice in the light theme (todo item for me, update the colors in light theme) and are not the most primair colors used in other dumps
- It displays both the main line number, and all line numbers of the excerpt. I think the main line number should be removed from the key. E.g:
  
  _Before_
  
  ```
  23. ...\ExceptionCasterTest.php:23: {
    22: {
    23:     return new \Exception('foo');
    24: }
  }
  ```
  
  _After_
  
  ```
  ...\ExceptionCasterTest.php:23: {
    22: {
    23:     return new \Exception('foo');
    24: }
  }
  ```

**Comment by nicolas-grekas** [comment]
Haha, the "line number" your removed is not a line number but an ordering number in the list :)

**Comment by nicolas-grekas** [comment]
Btw, are you OK with displaying the source code excerpt upfront or do you think it's better to require a click by default (I think the 2nd is better personalty)?

_IE_

capture du 2016-08-23 14-09-43

_VS_

capture du 2016-08-23 14-15-33

**Comment by nicolas-grekas** [comment]
@wouterj I'm done here, I invite you to reset --hard on it.
I enhanced a bit the Logs panel by making the "Message" column take 100%, and tweaking a few things.
With https://github.com/symfony/symfony/pull/19714, we should be done!

**Comment by wouterj** [comment]
I also like the 2nd one more.

I'll work on it later this afternoon. Thanks for your work!

**Comment by nicolas-grekas** [comment]
@wouterj rebased, perfect on my side :)

**Comment by wouterj** [comment]
@nicolas-grekas I've also implemented the security and forms panels now. I'm about to push all changes (I'm making screenshots at the moment).

**Comment by nicolas-grekas** [comment]
@wouterj commit added on top of yours, fixing the easiest comments I had and most importantly removing the added Twig node things. I figured out a way to not need them.
The comment about the bc break in FormDataExtractor still holds. Can you work on it?

**Comment by nicolas-grekas** [comment]
@wouterj I'll give it a try so don't hurry :)

**Comment by nicolas-grekas** [comment]
@wouterj self-comments addressed :)
Since I spent some time on cleaning the git history, please reset --hard (be careful not loosing any changes you made on this branch before):

```
git fetch https://github.com/nicolas-grekas/symfony.git profiler_dumper
git checkout profiler_dumper
git reset --hard FETCH_HEAD
```

**Comment by nicolas-grekas** [comment]
Oh, and should be green :)

**Comment by nicolas-grekas** [comment]
You'll notice that for FormDataExtractor, I copied it in a sub namespace and deprecated the existing one. Much easier to preserve BC this way.

**Comment by nicolas-grekas** [comment]
(looks like tests need a bit more love...)

**Comment by wouterj** [comment]
> (looks like tests need a bit more love...)

A lot more indeed. That's why I first want some opinions from core members before changing stuff again after fixing all tests.

> You'll notice that for FormDataExtractor, I copied it in a sub namespace and deprecated the existing one. Much easier to preserve BC this way.

I wonder if it's really needed to preserve BC here? It wasn't possible to change as the service didn't have any arguments in the Symfony framework.

**Comment by nicolas-grekas** [comment]
> if it's really needed to preserve BC here?

understood, you're right, I put it back were it was, and updated the constructor to fix the way the deprecation is triggered!

> opinions from core members before changing [...] tests

good idea.

OK to `reset --hard` now so that reviewers on your PR can see the latest patch?

**Comment by nicolas-grekas** [comment]
> +359 −354

:smiley:

**Comment by wouterj** [comment]
> OK to reset --hard now so that reviewers on your PR can see the latest patch?

Done

**Comment by nicolas-grekas** [comment]
@wouterj some tests fixed in latest commit.

**Comment by nicolas-grekas** [comment]
@wouterj All tests fixed, except one that is failing due to new code that is broken. I'll let you fix this remaining one since I didn't look at what it does  exactly.

**Comment by nicolas-grekas** [comment]
@wouterj @lyrixx, you might like the last commit:

capture du 2016-08-27 09-11-28

File: src/Symfony/Bundle/WebProfilerBundle/Twig/WebProfilerExtension.php [hunk 1]
```diff
@@ -88,6 +89,24 @@ public function dumpData(\Twig_Environment $env, Data $data, $maxDepth = 0)
...
+
+        if (false === strpos($message, '{')) {
+            return '<span class="log-message">'.$message.'</span>';
+        }
+
+        $replacements = array();
+        foreach ($context->getRawData()[1] as $k => $v) {
+            $v = '{'.twig_escape_filter($env, $k).'}';
+            $replacements['&quot;'.$v.'&quot;'] = $this->dumpData($env, $context->seek($k));
+            $replacements[$v] = $replacements['&quot;'.$v.'&quot;'];
```
Comment (wouterj) [inline src/Symfony/Bundle/WebProfilerBundle/Twig/WebProfilerExtension.php]: Wouldn't it be easier to do:

``` php
$replacements[$v] = $replacements['&quot;'.$v.'&quot;'] = $this->dumpData($env, $context->seek($k));
```

Comment (nicolas-grekas) [reply src/Symfony/Bundle/WebProfilerBundle/Twig/WebProfilerExtension.php]: Quote before because the order count, but yes

**Comment by wouterj** [comment]
Will look at the test failure tomorrow. The error cause needs a rewrite (probably using VarDumper's trace cloner). Problem here is that it's either an exception or a ConstraintViolation, is it possible to add a cloner for the ConstraintValidation trace as well?

**Comment by nicolas-grekas** [comment]
@wouterj I managed to figure out how to fix FormDataExtractor.
Tests should be green now.
WDYT, esp. concerning form errors?

**Comment by nicolas-grekas** [comment]
Just pushed a small update

**Comment by nicolas-grekas** [comment]
@wouterj could you please check and reset --hard again if the diff is ok to you?
Should be reviewable by deciders on my side!

**Comment by wouterj** [comment]
@nicolas-grekas problem with the form error causes is that it's now displaying `getPrevious()`/`getCause()` already when expanding the first dump. So there probably is no need looping through these?

**Comment by nicolas-grekas** [comment]
You mean that cause/previous are duplicated: once recursively, then linearly?
I thing we should keep the linear version as it makes spotting errors more straightforward.

**Comment by nicolas-grekas** [comment]
I just removed "previous" from exception dumps.
Causes where already removed, isn't it?
Another comment?

File: src/Symfony/Component/Form/Extension/DataCollector/FormDataExtractor.php [hunk 2]
```diff
@@ -199,4 +200,34 @@ private function buildId(FormInterface $form)
...
+     *
+     * @return Data
+     */
+    private function cloneVar($var)
+    {
+        if (null === $this->cloner) {
+            $this->cloner = new VarCloner();
+            $this->cloner->addCasters(array(
+                \Exception::class => function (\Exception $e, array $a, Stub $s) { unset($a["\0Exception\0previous"]); ++$s->cut; return $a; },
+                FormInterface::class => function (FormInterface $f, array $a, Stub $s) { return StubCaster::cutInternals($f, $a, $s, true); },
```
Comment (wouterj) [inline src/Symfony/Component/Form/Extension/DataCollector/FormDataExtractor.php]: does this line avoid dumping the Form in the constraint violations? If it does, I'm not sure if we should include `root` at all, it only shows `Form {#986 …13}` now, not that usefull.

**Comment by wouterj** [comment]
upstream PR updated

File: src/Symfony/Component/Form/Extension/DataCollector/FormDataExtractor.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Form/Extension/DataCollector/FormDataExtractor.php]: You're right.
Would it be more useful to display the form name in front of this "root" key? Any better idea?

**Comment by nicolas-grekas** [comment]
Form name on stage :)

File: src/Symfony/Component/Form/Extension/DataCollector/FormDataExtractor.php (hunk 2 above)
Comment (wouterj) [reply src/Symfony/Component/Form/Extension/DataCollector/FormDataExtractor.php]: I think name and type are the most usefull yeah. (I quite like these little caster closures)

Comment (wouterj) [reply src/Symfony/Component/Form/Extension/DataCollector/FormDataExtractor.php]: applied this in the upstream PR

**Comment by nicolas-grekas** [comment]
PR updated: I cherry-picked your changes, fixed tests and fabbot. Should be green :)

File: src/Symfony/Component/Form/Extension/DataCollector/FormDataExtractor.php [hunk 3]
```diff
@@ -199,4 +198,43 @@ private function buildId(FormInterface $form)
...
+                        unset($a[$k]);
+                        ++$s->cut;
+                    }
+
+                    return $a;
+                },
+                FormInterface::class => function (FormInterface $f, array $a, Stub $s) {
+                    return array(
+                        Caster::PREFIX_VIRTUAL.'name' => $f->getName(),
+                        Caster::PREFIX_VIRTUAL.'type_class' => get_class($f->getConfig()->getType()->getInnerType()),
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/DataCollector/FormDataExtractor.php]: renamed this to type_class to be consistent with https://github.com/nicolas-grekas/symfony/pull/6/files#diff-1b203840c96e2f4a2f0386788e2765d3R53

File: src/Symfony/Component/Form/Tests/Extension/DataCollector/FormDataExtractorTest.php [hunk 4]
```diff
@@ -394,13 +397,35 @@ public function testExtractViewVariables()
...
+    {
+        $h = fopen('php://memory', 'r+b');
+        $dumper = new CliDumper($h);
+        $inlined = array();
+
+        foreach ($extraction as $k => $v) {
+            if (is_array($v)) {
+                $inlined[$k] = $this->inlineData($v);
+            } elseif ($v instanceof Data) {
+                $dumper->dump($v->withRefHandles(false));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Tests/Extension/DataCollector/FormDataExtractorTest.php]: We can't use `true` as second arg without bumping the minimum required version. Not worth it.

**Comment by nicolas-grekas** [comment]
@wouterj I rebased on latest master so you can see how it goes with latests vardumper features :)

**Comment by nicolas-grekas** [comment]
@wouterj I just fixed another glitch. Can you please sync your PR again and notify others that this is ready?

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/form.html.twig [hunk 5]
```diff
@@ -442,7 +439,7 @@
             {% endif %}
 
             <span {% if has_error or data.has_children_error|default(false) %}class="has-error"{% endif %}>
-                {{ name|default('(no name)') }} {% if data.type_class is defined %}[<abbr title="{{ data.type_class }}">{{ data.type_class|split('\\')|last }}</abbr>]{% endif %}
+                {{ name|default('(no name)') }} [{{ profiler_dump(data.type_class) }}]
```
Comment (wouterj) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/form.html.twig]: as dumps are block elements and not inline, this will completely destroy the layout of the tree and page (same applies to line 460):

sf-profiler-dumper-form-bug

We should either allow to dump inline (`<code>...</code>` instead of `<pre>...</pre>`) or we should revert this change

Comment (nicolas-grekas) [reply src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/form.html.twig]: Indeed, I missed a few CSS here. I did something similar for log contexts, they are inline, while still in `<pre>`, we should do the same here also.

Comment (wouterj) [reply src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/form.html.twig]: ah, I see. I'll take care of it

**Comment by nicolas-grekas** [comment]
ping @wouterj,
var-dumper is now an optional dep of http-kernel again.
Any blocker to submit our work to reviewers?
